src/rdkafka_transport.c

```diff
@@ -877,6 +1014,147 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
 		}
 	}
 
+    /* Register the peer verify callback on the context */
+    if (rk->rk_conf.ssl.cert_verify_cb) {
+
+        unsigned int mode = SSL_VERIFY_PEER | SSL_VERIFY_CLIENT_ONCE | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+
+        rd_kafka_dbg(rk, SECURITY, "SSL",
+            "Setting verification mode %u", mode);
+
+        SSL_CTX_set_verify(
+            ctx,mode,
+            NULL);
+
+        rd_kafka_dbg(rk, SECURITY, "SSL",
+            "Registering verification callback");
+
+        SSL_CTX_set_cert_verify_callback(
+            ctx,
+            verify_broker_callback,
+            rk);
+    }
+
+    if (rk->rk_conf.ssl.cert_retrieve_cb) {
+
+        unsigned char* buffer = NULL;
+        long len = 0;
+
+        rd_kafka_dbg(rk, SECURITY, "SSL",
+            "Calling callback to retrieve client's public key certificate");
+        len = rk->rk_conf.ssl.cert_retrieve_cb(RD_KAFKA_CERTIFICATE_PUBLIC_KEY, &buffer, rk->rk_conf.opaque);
```

Comment:
pass an errstr + size so that we can have the user give us a meaningful error message when this fails.
Should return -1 on error instead of 0.